### PR TITLE
Opencl instrumentation fixgtpinbindings

### PR DESCRIPTION
**Pull request by jmellorcrummey**
- bind gtpin routines only if they are to be used. call them through function pointers rather than directly
- add support for replacing dlmopen with dlopen and loading all shared libraries into the initial namespace
- in hpcrun, replace dlmopen with dlopen when gtpin is used
- in hpcrun, add --dlopen and --dlmopen arguments to force the handling of choice just in case
- ensure that device finalizers never wait for event completions that may never occur when an application performs anything other than a normal exit

### CODE REVIEW

**Review by blue42u — APPROVED** [review]
Assuming it compiles and runs as expected, looks good. The below are just pedantic comments.

File: src/tool/hpcrun/gpu/opencl/opencl-api.c [hunk 1]
```diff
@@ -866,59 +901,14 @@ opencl_bind
...
-
-      // TO-DO: AARON using malloc instead of hpcrun_malloc gives extra garbage characters in file name
-      char *filename = (char *)hpcrun_malloc(sizeof(fileno) + 1);
-      *filename = fileno + '\0';
-      f_ptr = fopen(filename, "w");
-      fwrite(strings[i], lengths[i], 1, f_ptr);
-    }
-    fclose(f_ptr);
+  int status = opencl_bind_once();
+  if (status == 0) {
```
[the remark this answers is not in the document]
Comment (jmellorcrummey) [reply src/tool/hpcrun/gpu/opencl/opencl-api.c:905]: I was lazy here. 

Actually, it needs more reworking than what you suggested. The binding macros are used in several places in HPCToolkit. The status codes should be renamed BIND_SUCCESS and BIND_FAILURE and checked consistently. (The binding macros used in opencl_bind_once() are used for ROCm, CUDA, and CUPTI bindings as well.) I'll fix those too.

File: src/tool/hpcrun/gpu/opencl/opencl-api.c [hunk 2]
```diff
@@ -650,6 +634,72 @@ opencl_wait_for_all_pending_operations
...
+{
+  pthread_once(&is_initialized, opencl_bind_once_helper);
+  return opencl_bind_result;
+}
+
+
+static bool
+opencl_noaction
+(
+ char *name
```
[the remark this answers is not in the document]
Comment (jmellorcrummey) [reply src/tool/hpcrun/gpu/opencl/opencl-api.c:691]: I passed the names to print debugging messages but then managed to debug it before adding the code to print. All of the opencl_noaction() calls will be eliminated with the planned new preloaded interface for injecting wrappers only when they are needed.

After I add the new wrapping library, this code will disappear.

File: src/tool/hpcrun/scripts/hpcrun.in [hunk 3]
```diff
@@ -137,6 +138,14 @@ Options: Profiling (Defaults shown in curly brackets {})
...
                        hpctoolkit_sampling_start().
 
+  ---dlmopen
+                       Allow dlmopen to create alternate namespaces.
+
+  --dlopen
+                       Use dlopen in place of dlmopen, which has the effect of 
+                       loading all shared libraries into the application's 
+                       namespace. (Caution: this may crash your program!)
+
```
[the remark this answers is not in the document]
Comment (jmellorcrummey) [reply src/tool/hpcrun/scripts/hpcrun.in:148]: I admit my names were cryptic. I like your suggestion above. I will move these flags into a section for "Flags for expert use only".

File: src/tool/hpcrun/scripts/hpcrun.in [hunk 4]
```diff
@@ -340,6 +359,20 @@ do
 	    export HPCRUN_OMP_SERIAL_ONLY=1
 	    ;;
 
+
+	# --------------------------------------------------
+
+	--dlopen )
+	    dlopen=yes
```
[the remark this answers is not in the document]
Comment (jmellorcrummey) [reply src/tool/hpcrun/scripts/hpcrun.in:366]: I will add this.

**Review by mxz297 — APPROVED** [review]
@jmellorcrummey Looks good. Based on the existing comments, we will first merge this PR and then do the refactoring to move opencl and level 0 function overrides to separate  libraries?

**Comment by Jokeren** [comment]
@mxz297 We found another problem on Friday. For some dpcpp applications, hpcrun records empty gpubins. I haven't investigated the problem yet.

File: src/tool/hpcrun/gpu/opencl/opencl-api.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (jmellorcrummey) [reply src/tool/hpcrun/gpu/opencl/opencl-api.c:904]: the complete overhaul of return codes is done for managing dynamic binding. our libdl.h file now defines
DYNAMIC_BINDING_STATUS_{UNINIT,ERROR,OK}. All constants have been replaced by symbolic values.

File: src/tool/hpcrun/gpu/opencl/opencl-api.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (jmellorcrummey) [reply src/tool/hpcrun/gpu/opencl/opencl-api.c:690]: this code stays temporarily until the next refactoring, which will add a separate library that wraps the OpenCL interface.

File: src/tool/hpcrun/scripts/hpcrun.in (hunk 3 above)
[the remark this answers is not in the document]
Comment (jmellorcrummey) [reply src/tool/hpcrun/scripts/hpcrun.in]: I overhauled the options, collecting options that should only be considered when hpcrun crashes an application.

The new options are
--namespace-multiple
--namespace-single

File: src/tool/hpcrun/scripts/hpcrun.in [hunk 5]
```diff
@@ -340,6 +359,20 @@ do
...
+	    ;;
+
+
+	# --------------------------------------------------
+
+	--dlmopen )
+	    dlmopen=yes
```
[the remark this answers is not in the document]
Comment (jmellorcrummey) [reply src/tool/hpcrun/scripts/hpcrun.in]: done

File: src/tool/hpcrun/scripts/hpcrun.in [hunk 6]
```diff
@@ -421,6 +454,16 @@ if test -z "$1" ; then
...
 fi
 
+
+# replace dlmopen with dlopen as directed
+if test -z "$dlmopen" ; then
+  case "$dlopen" in
+      '' ) ;;
+      yes ) preload_list="${hpcrun_dir}/libhpcrun_dlmopen.so ${preload_list}"
+  esac
+fi
```
[the remark this answers is not in the document]
Comment (jmellorcrummey) [reply src/tool/hpcrun/scripts/hpcrun.in]: argument and option processing rewritten

File: src/tool/hpcrun/scripts/hpcrun.in [hunk 7]
```diff
@@ -214,6 +223,9 @@ arg_ok()
 
 preload_list=
 roctracer_lib=
+gtpin_lib=
+dlmopen=
+dlopen=
```
[the remark this answers is not in the document]
Comment (jmellorcrummey) [reply src/tool/hpcrun/scripts/hpcrun.in]: argument processing rewritten

File: src/tool/hpcrun/scripts/hpcrun.in (hunk 4 above)
[the remark this answers is not in the document]
Comment (jmellorcrummey) [reply src/tool/hpcrun/scripts/hpcrun.in]: done.

File: src/tool/hpcrun/scripts/hpcrun.in [hunk 8]
```diff
@@ -277,6 +289,13 @@ do
 		MPI* )     preload_list="${preload_list} ${hpcrun_dir}/libhpcrun_mpi.so" ;;
 		gpu=amd) roctracer_lib="${roctracer_lib_path}"
 			 export HIP_ENABLE_DEFERRED_LOADING=0;;
+
+		gpu=opencl,inst) gtpin_lib="${gtpin_lib_path}"
+		    dlopen=yes;;
+
+		gpu=level0,inst) gtpin_lib="${gtpin_lib_path}"
+		    dlopen=yes;;
```
[the remark this answers is not in the document]
Comment (jmellorcrummey) [reply src/tool/hpcrun/scripts/hpcrun.in]: analogous change made.

**Comment by jmellorcrummey** [comment]
Completed modifications according to comments. Also, modified hpcrun script to not introduce ::.

**Pull request merged**